### PR TITLE
Install LTTng and related packages on Linux

### PR DESCRIPTION
**Pull request by christophebedard**
Part of https://github.com/ros2/ros2/issues/1177

See https://github.com/ros2/ros2_tracing/pull/31

Signed-off-by: Christophe Bedard <christophe.bedard@apex.ai>

### CODE REVIEW

File: linux_docker_resources/Dockerfile-RHEL
Comment (christophebedard) [inline linux_docker_resources/Dockerfile-RHEL:137]: I think `lttng-ust` should work, but I'm not completely sure about `python3-lttng`.

Comment (cottsay) [reply linux_docker_resources/Dockerfile-RHEL:137]: RHEL 8 has `python3-lttng`, but the problem is that it's too old and tracetools breaks if it's present. I removed it in #657. Until tracetools handles the old version gracefully (which I'd still very much like to see), this is the best workaround.

**Comment by clalancette** [comment]
> I can do `${UBUNTU_DISTRO} = jammy` and `${ROS_DISTRO} != humble` for Ubuntu and `${EL_RELEASE} != 8` for RHEL for all these new packages.

Yeah, that is probably a good idea.  I'll also suggest putting this in a separate block in the Ubuntu Dockerfile, similar to https://github.com/ros2/ci/blob/7cbb9f6d57c41cac15fa8fd2b8020d2f7fe388de/linux_docker_resources/Dockerfile#L97

**Comment by christophebedard** [comment]
> I'll also suggest putting this in a separate block in the Ubuntu Dockerfile

Done!

**Review by clalancette — APPROVED** [review]
This looks good to me, though I'll want to see Linux CI for all of Foxy, Humble, and Rolling on both Ubuntu and RHEL.  I'd also like to get one more approval from @cottsay .

File: linux_docker_resources/Dockerfile-RHEL
Comment (cottsay) [inline linux_docker_resources/Dockerfile-RHEL:116]: As the Debian side is getting the development files for `lttng-ust`, the RHEL side should as well.
```suggestion
    $(if test ${EL_RELEASE} != 8; then echo lttng-ust-devel; fi) \
```

The rosdep DB already reflects this properly: https://github.com/ros/rosdistro/blob/64bff7eb4ad6e18e0374219abb252a37462a94ae/rosdep/base.yaml#L4161

**Review by cottsay — CHANGES_REQUESTED** [review]
One quick change

**Review by cottsay — APPROVED** [review]
Thanks!

**Comment by clalancette** [comment]
OK, I've pushed an exact copy of this branch to the origin.  Here are some CI runs with that in place:

Foxy:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18352/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12881/)
* RHEL-8 [Build Status](http://ci.ros2.org/job/ci_linux-rhel/353/)

Humble:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18353/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12882/)
* RHEL-8 [Build Status](http://ci.ros2.org/job/ci_linux-rhel/354/)

Rolling:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18354/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12883/)
* RHEL-9 [Build Status](http://ci.ros2.org/job/ci_linux-rhel/355/)

**Comment by christophebedard** [comment]
I thought we didn't need `lttng-tools` (which provides the CLI), but I forgot that I added it as an `<exec_depend>` of `tracetools` and that we need it in `tracetools_trace`: https://github.com/ros2/ros2_tracing/blob/e664e5bbd28b42da6430623fc1bcf644d6b6a449/tracetools_trace/tracetools_trace/tools/lttng.py#L90-L94. I'll add it as exec and test dependency of `tracetools_trace` in https://github.com/ros2/ros2_tracing/pull/31 and add it here.

**Review by cottsay — APPROVED** [review]

**Comment by christophebedard** [comment]
I also noticed that there was no rosdep rule for `lttng-tools` for RHEL, so I added it: https://github.com/ros/rosdistro/pull/36641

**Comment by clalancette** [comment]
All right.  I'm only going to run the latest change against Rolling again, given that it is protected under the same checks as the previous code.

**Comment by clalancette** [comment]
Rolling:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18357/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12887/)
* RHEL-9 [Build Status](http://ci.ros2.org/job/ci_linux-rhel/356/)

**Comment by christophebedard** [comment]
The test failures on RHEL don't seem to be related, but I can't find the same failures in recent jobs.

**Comment by clalancette** [comment]
> The test failures on RHEL don't seem to be related, but I can't find the same failures in recent jobs.

That one is a flaky test; you can see a failure of it in https://ci.ros2.org/view/nightly/job/nightly_linux-rhel_repeated/1448/#showFailuresLink .  I really don't think it is involved here at all.

So I think we are good with this change.

Once we merge this in, that means that we'll be building tracepoints in by default for Rolling in our nightly CI jobs and our nightly packaging (tarball) jobs.  What needs to happen to have it enabled in the debians?  Do we just need to do a release of `ros2_tracing` for that to happen?

**Comment by christophebedard** [comment]
> Once we merge this in, that means that we'll be building tracepoints in by default for Rolling in our nightly CI jobs and our nightly packaging (tarball) jobs. What needs to happen to have it enabled in the debians? Do we just need to do a release of `ros2_tracing` for that to happen?

Specifically after merging https://github.com/ros2/ros2_tracing/pull/31 (which I think we should merge right after this), yes! LTTng will be picked up as a dependency, and the tracepoints will be compiled in.

Well, that's what I think will happen, anyway. I can check the testing repo and validate once it's re-built.

**Comment by christophebedard** [comment]
Next step before merging would be to run CI using this branch and https://github.com/ros2/ros2_tracing/pull/31.

**Comment by clalancette** [comment]
> Next step before merging would be to run CI using this branch and [ros2/ros2_tracing#31](https://github.com/ros2/ros2_tracing/pull/31). See [ros2/ros2_tracing#31 (comment)](https://github.com/ros2/ros2_tracing/pull/31#issuecomment-1489374096)

Awesome, the jobs there look good.

So what is the next step here?  Do you want to merge the ros2_tracing PR, then I can merge this one?

**Comment by christophebedard** [comment]
> So what is the next step here? Do you want to merge the ros2_tracing PR, then I can merge this one?

Since packages will now depend on LTTng, and since CI doesn't seem to run `rosdep install` (except for RHEL, it seems), we should merge this first. Then I'll merge https://github.com/ros2/ros2_tracing/pull/31 right after.

**Comment by clalancette** [comment]
> Since packages will now depend on LTTng, and since CI doesn't seem to run `rosdep install` (except for RHEL, it seems), we should merge this first. Then I'll merge [ros2/ros2_tracing#31](https://github.com/ros2/ros2_tracing/pull/31) right after.

OK, got it.

So, the only thing is that I'm on PTO, not checking email, starting about now until Monday (April 3).  Despite the fact that I think we've done sufficient testing here, I'm somewhat reluctant to merge something like this and disappear.  So either we should have @cottsay merge it and keep an eye on it tomorrow, or I can do it first thing Monday morning when I'm back.  @cottsay what do you think?

**Comment by christophebedard** [comment]
Waiting until Monday as a precaution is fine with me.

**Pull request merged**
